### PR TITLE
Issue fixed in nncf/common/sparsity- [NNCF] Make NNCF common sparsity code pass mypy checks #22312

### PR DESCRIPTION
**Pull request by AishwaryaDekhane**
[NNCF#22312.pdf](https://github.com/openvinotoolkit/nncf/files/14056617/NNCF.22312.pdf)
### Changes

Improving the type safety and maintainability of NNCF's common sparsity code (nncf/common/sparsity) using mypy checks. Mypy is a static type checker that helps identify potential errors before running the code.
-Missing or incorrect type annotations on functions and variables, making it difficult for mypy to understand the expected data types.
-Functions returning incompatible types, causing mismatches and potential errors.
-Accessing non-existent attributes in classes, leading to runtime errors. 
-Using untyped functions within typed contexts, introducing uncertainty about data types.

The implemented fixes involved adding comprehensive type annotations throughout the code, correcting return types, addressing attribute access issues, and ensuring consistent type usage. This improves the overall code quality and makes it easier to understand and maintain.

### CODE REVIEW

**Comment by MaximProshin** [comment]
@alexsu52 @andrey-churkin fyi

File: nncf/api/compression.py
Comment (vshampor) [inline nncf/api/compression.py:36]: I understand you need to fix this error:
`nncf/common/sparsity/schedulers.py:273: error: "CompressionLoss" has no attribute "current_sparsity"  [attr-defined]`
but this is not the right way to do it since this exact class is a) a pure interface, and is not supposed to have `__init__`, and b) even if `self.current_sparsity` were made a property, `CompressionLoss` would not be the right place for it since the `current_sparsity` is only relevant to sparsity-related loss use cases and not for the general ABC such as `CompressionLoss`.

So this change most definitely cannot be accepted, but we need to think how to correct the situation. Analyzing the original problem at nncf/common/sparsity/schedulers.py:273 it now seems somewhat strange to me that the current sparsity value should really be accessible through the `SparsityController.scheduler.loss` field and not, say, through the `SparsityController` itself. Another solution would be to make sure that the `SparsityController` class definition types the `loss` field/property to have the `current_sparsity` attribute, i.e. by typing the `loss` to be something like `SparsityLoss`. 

@andrey-churkin what do you think?

**Review by vshampor — CHANGES_REQUESTED** [review]
Greetings! Thank you for the contribution!

I can see from the PR that you've got the task right and I checked that the mypy indeed shows no errors in the specified set of files. However, we still need the precommit checks to pass to ensure the logic was not impacted. I can see right now, though, that the logic will be impacted and the precommit checks would likely fail.

I've left some comments for you to address, please check them out. There is one comment about `current_sparsity` for which we devs will need to make a decision first, so you can in the meanwhile concentrate on the other comments.

I see that you have not used the auto code formatting for the PR - our CI checks for this and will fail if the formatting is incorrect. Please check this section of the guide out:
https://github.com/openvinotoolkit/nncf/blob/develop/docs/styleguide/PyGuide.md#2-automating-code-formatting

Basically you need to run `make pre-commit` and ensure that it is passing before submitting the PR state. Note that the `make pre-commit` command may make in-place changes to the files (to autoformat etc.), so don't forget to add these changes to the git commit after running the command.

**Comment by AishwaryaDekhane** [comment]
I'm so sorry about the pre-commit checks and formatting issues! Thank you so much for your thorough review and clear instructions. I completely understand your concerns about logic impact and ensuring everything adheres to the guidelines.
I'll definitely address all your comments in the PR, focusing on the ones I can tackle myself for now, and wait for guidance on current_sparsity. Additionally, I'll run make pre-commit and utilize the formatting guide to ensure everything passes the CI checks before resubmitting the PR.
Thank you again for your patience and valuable feedback. I'm truly grateful for your time and support, and I'll strive to deliver a PR that meets all the expectations.
I'll get back to you soon with the updated PR once I've implemented the necessary changes. Thank you.

**Comment by AishwaryaDekhane** [comment]
I have implemented the aforementioned changes. Can you please verify them?

**Comment by AishwaryaDekhane** [comment]
pre-commit

File: nncf/common/sparsity/schedulers.py
Comment (AishwaryaDekhane) [reply nncf/common/sparsity/schedulers.py]: Resolved

File: nncf/common/sparsity/schedulers.py
Comment (AishwaryaDekhane) [reply nncf/common/sparsity/schedulers.py]: Resolved

File: nncf/common/sparsity/schedulers.py
Comment (AishwaryaDekhane) [reply nncf/common/sparsity/schedulers.py]: Resolved

**Comment by rkazants** [comment]
@vshampor, @alexsu52, please help with review and run CI for this PR.

Best regards,
Roman

**Review by vshampor — COMMENTED** [review]
Thanks for the contribution. In its current state the PR still won't work with the existing unit tests and logic. I have pushed a commit that will hopefully fix the rest of the inconsistencies, will see if the CI is OK. I have left a couple of comments on your previously suggested code, JFYI.

**Review by vshampor — APPROVED** [review]

File: nncf/common/sparsity/controller.py
Comment (vshampor) [inline nncf/common/sparsity/controller.py:39]: @alexsu52 this exposes a new property on the derived `nncf.torch.sparsity.base_algo.BaseSparsityAlgoController` which is marked as API, hence the API marker was (correctly) assigned to this PR. This seems to be a good property for the sparsity controllers, and helps resolve one of the original mypy issues with our code. Do you have any comments or objections for that?

**Comment by vshampor** [comment]
Looks like now this is mergeable. Waiting on review from @alexsu52 and @andrey-churkin

**Review by andrey-churkin — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]
Thanks for the contribution!

**Pull request merged**
